### PR TITLE
[IA-4653] Fix region logic

### PR DESCRIPTION
**Pull request by jdcanas**
[lily's note] Some runtime responses don't have a region specified; this causes the Cloud Environments page to break. Normalize the region field on all runtime configs in the data layer so that the rest of the UI can trust the value to be truthy.

### CODE REVIEW

File: src/analysis/utils/runtime-utils.ts [hunk 1]
```diff
@@ -48,6 +48,7 @@ export const defaultLocation = 'US-CENTRAL1';
 
 export const defaultComputeZone = 'US-CENTRAL1-A';
 export const defaultComputeRegion = 'US-CENTRAL1';
+export const defaultAzureComputeRegion = 'eastus';
```
Comment (LizBaldo) [inline src/analysis/utils/runtime-utils.ts]: I am pretty sure that our default region for Azure is southcentral now, better double check with the dsp-azure channel

Comment (cindy-broadinstitute) [reply src/analysis/utils/runtime-utils.ts]: Yes, was going to say that it is 'southcentral'

Comment (lucymcnatt) [reply src/analysis/utils/runtime-utils.ts]: at least with all my current Azure workspaces, the region is `southcentralus`

Comment (mlilynolting) [reply src/analysis/utils/runtime-utils.ts]: this may duplicate `defaultAzureRegion` at https://github.com/DataBiosphere/terra-ui/blob/07545d8402606b376ea5605cc01da27faf7711ff/src/libs/azure-utils.js#L10C39-L10C39 - tho that value should probably change to "southcentralus"

File: src/analysis/utils/runtime-utils.ts [hunk 2]
```diff
@@ -74,15 +75,15 @@ export type NormalizedComputeRegion = NominalType<string, 'ComputeRegion'>;
 // TODO: test when zone and region have types
 export const getNormalizedComputeRegion = (config: GoogleRuntimeConfig | AzureConfig): NormalizedComputeRegion => {
   if (isGceConfig(config) || isGceWithPdConfig(config)) {
-    return getRegionFromZone(config.zone).toUpperCase() as NormalizedComputeRegion;
+    return (getRegionFromZone(config.zone) || defaultComputeRegion).toUpperCase() as NormalizedComputeRegion;
```
Comment (mlilynolting) [inline src/analysis/utils/runtime-utils.ts]: `GceConfig` and `GceWithPdConfig` both have `zone` as nonoptional. Are you concerned about a case where it's coming in as an empty string?

File: src/analysis/utils/runtime-utils.ts [hunk 3]
```diff
@@ -74,15 +75,15 @@ export type NormalizedComputeRegion = NominalType<string, 'ComputeRegion'>;
...
   }
   if (isDataprocConfig(config)) {
-    return config.region.toUpperCase() as NormalizedComputeRegion;
+    return (config.region || defaultComputeRegion).toUpperCase() as NormalizedComputeRegion;
```
Comment (mlilynolting) [inline src/analysis/utils/runtime-utils.ts]: same here; seems like `region` is nonoptional anyway

File: src/analysis/utils/runtime-utils.ts [hunk 4]
```diff
@@ -74,15 +75,15 @@ export type NormalizedComputeRegion = NominalType<string, 'ComputeRegion'>;
...
   }
   if (isAzureConfig(config)) {
-    return config.region.toUpperCase() as NormalizedComputeRegion;
+    return (config.region || defaultAzureComputeRegion).toUpperCase() as NormalizedComputeRegion;
```
Comment (mlilynolting) [inline src/analysis/utils/runtime-utils.ts:84]: is this method ever used in an Azure context? All I can find is Environments.ts, and that's specifically for showing runtime region.

File: src/analysis/utils/runtime-utils.ts (hunk 3 above)
Comment (jdcanas) [reply src/analysis/utils/runtime-utils.ts]: This code takes a `RuntimeConfig` which can be one of 4 concrete types, two of which have zone and two of which have region, and normalizes it to a region for both cost calculation and display.

File: src/analysis/utils/runtime-utils.ts (hunk 1 above)
Comment (jdcanas) [reply src/analysis/utils/runtime-utils.ts]: Hmm... so based on this discussion I'm going to remove putting the default region in this function.

The logic to put it here was initially flawed: if leo isn't returning a region, we should be putting 'unknown' here instead of a point-in-time default value.

It's a good safety mechanism to protect us from the fact this field is optional sometimes, but leo is trying to fill it in everywhere

File: src/analysis/utils/runtime-utils.ts [hunk 5]
```diff
@@ -73,16 +73,17 @@ export type NormalizedComputeRegion = NominalType<string, 'ComputeRegion'>;
 
 // TODO: test when zone and region have types
 export const getNormalizedComputeRegion = (config: GoogleRuntimeConfig | AzureConfig): NormalizedComputeRegion => {
+  const regionNotFoundPlaceholder = 'Unknown';
```
Comment (mlilynolting) [inline src/analysis/utils/runtime-utils.ts]: sorry for the trail of comments, trying to keep up - I'm definitely missing something, because `region` looks like a required field for all of these `*Config` types. How is a null value getting into this method? Or are we guarding against the empty String case?

**Review by mlilynolting — COMMENTED** [review]
I need an explainer :(

File: src/analysis/utils/runtime-utils.ts (hunk 5 above)
Comment (jdcanas) [reply src/analysis/utils/runtime-utils.ts]: It is optional here: 
image

Comment (jdcanas) [reply src/analysis/utils/runtime-utils.ts]: Should probably be added to the UI model...

File: src/analysis/utils/runtime-utils.ts (hunk 3 above)
Comment (mlilynolting) [reply src/analysis/utils/runtime-utils.ts]: yes, but each of the specific checks is inside of a typeguard conditional like `isDataprocConfig(config)`, and knows which type it's working with.

File: src/analysis/utils/runtime-utils.ts (hunk 5 above)
Comment (mlilynolting) [reply src/analysis/utils/runtime-utils.ts]: Is the UI type just not enforced then? Shouldn't Typescript be complaining? Or does it not bc some of the pipeline is still Javascript?

Comment (mlilynolting) [reply src/analysis/utils/runtime-utils.ts]: If we do expect a non-null value, we should normalize it in the data layer, not in the utils logic, right?

File: src/libs/ajax/leonardo/models/runtime-config-models.ts [hunk 6]
```diff
@@ -73,3 +75,7 @@ export const isGceConfig = (config: RuntimeConfig): config is GceConfig => {
 };
 
 export const isAzureConfig = (config: RuntimeConfig): config is AzureConfig => config.cloudService === 'AZURE_VM';
+
+export type NormalizedRuntimeConfig = {
```
Comment (mlilynolting) [inline src/libs/ajax/leonardo/models/runtime-config-models.ts]: Would it be possible to slip `normalizedRegion` into the base `RuntimeConfig`, so that `RuntimeConfig` is just always normalized? Or are there other factors preventing that?

Comment (jdcanas) [reply src/libs/ajax/leonardo/models/runtime-config-models.ts]: I don't want to mutate the original return from the API, only add to it in clear ways, so we don't obfuscate the data itself

Comment (mlilynolting) [reply src/libs/ajax/leonardo/models/runtime-config-models.ts]: Adding a field is a mutation but not a lossy one. Is that much different?

Comment (jdcanas) [reply src/libs/ajax/leonardo/models/runtime-config-models.ts]: Yes, a mutation that is not a lossy one is a decoration, and thus the model is backwards compatible. In other words, a runtimeConfig that is decorated can be used in the code as a runtime config that is not decorated, but the reverse isn't true.

Comment (mlilynolting) [reply src/libs/ajax/leonardo/models/runtime-config-models.ts]: So isn't it pretty safe to decorate RuntimeConfig directly? Like, rename existing RuntimeConfig to RawRuntimeConfig, and then make a new RuntimeConfig which is the decorated version, for use everywhere except the API layer?

File: src/analysis/utils/runtime-utils.ts (hunk 2 above)
Comment (jdcanas) [reply src/analysis/utils/runtime-utils.ts]: resolved in person, addressed the model discrepancies and put defaults where necessary

File: src/analysis/utils/runtime-utils.ts (hunk 5 above)
Comment (jdcanas) [reply src/analysis/utils/runtime-utils.ts]: Resolved in person, it is now normalized in the data layer

File: src/libs/ajax/leonardo/models/runtime-config-models.ts (hunk 6 above)
Comment (jdcanas) [reply src/libs/ajax/leonardo/models/runtime-config-models.ts]: We came up with a solution that provided stronger type safety and backwards compatibility, see the updated version of this file

File: src/libs/ajax/leonardo/Runtimes.ts [hunk 7]
```diff
@@ -31,14 +46,34 @@ const isAzureRuntimeWrapper = (obj: any): obj is AzureRuntimeWrapper => {
   return castObj && !!castObj.workspaceId && !!castObj.runtimeName;
 };
 
+const getNormalizedComputeConfig = (config: RawRuntimeConfig): RuntimeConfig => ({
+  ...config,
+  normalizedRegion: getNormalizedComputeRegion(config),
+});
+
+const getNormalizedComputeRegion = (config: RawRuntimeConfig): NormalizedComputeRegion => {
```
Comment (jdcanas) [inline src/libs/ajax/leonardo/Runtimes.ts:54]: note that this is now a private method, and methods dealing with the raw config are not exposed in favor of only exposing the sanitized object.

File: src/analysis/ContextBar.test.ts [hunk 8]
```diff
@@ -239,7 +240,7 @@ const cromwellOnAzureRunning: App = {
   region: 'us-central1',
 };
 
-const rstudioRuntime: Runtime = {
+const rstudioRuntime: ListRuntimeItem = {
```
Comment (mlilynolting) [inline src/analysis/ContextBar.test.ts:243]: These could still be Runtime objects, right? Don't Runtimes now have the normalized region in their configs?

File: src/analysis/utils/cost-utils.ts [hunk 9]
```diff
@@ -80,9 +79,9 @@ export const getGpuCost = (gpuType: string, numGpus: number, region: string): nu
 // This function deals with runtimes that are paused
 // All disks referenced in this function are boot disks (aka the disk google needs to provision by default for OS storage)
 // The user pd cost for a runtime is calculated elsewhere
-export const runtimeConfigBaseCost = (config: GoogleRuntimeConfig): number => {
+export const runtimeConfigBaseCost = (config: RuntimeConfig): number => {
```
Comment (mlilynolting) [inline src/analysis/utils/cost-utils.ts]: I think this can stay a GoogleRuntimeConfig. The scope of this method is GCP-only, right?

File: src/libs/ajax/leonardo/Runtimes.ts [hunk 10]
```diff
@@ -186,13 +224,14 @@ export const Runtimes = (signal: AbortSignal) => {
       // [IA-3710] In order to keep the front-end backwards compatible, any Azure tool labels
       // will be changed to JupyterLab.
       const runtimeList = await res.json();
-      const runtimesWithToolLabelDecorated = _.map((runtime) => {
+      const runtimesWithToolLabelDecorated: RawListRuntimeItem[] = _.map((runtime) => {
```
Comment (mlilynolting) [inline src/libs/ajax/leonardo/Runtimes.ts:227]: This could, I think, become something we move into `getNormalizedListRuntime`. But not in this PR

File: src/libs/ajax/leonardo/Runtimes.ts [hunk 11]
```diff
@@ -203,7 +242,9 @@ export const Runtimes = (signal: AbortSignal) => {
         `api/v2/runtimes/${workspaceId}?${qs.stringify({ saturnAutoCreated: true, ...labels })}`,
         _.mergeAll([authOpts(), appIdentifier, { signal }])
       );
-      return res.json();
+      const runtimes = await res.json();
+      const normalizedRuntimes = _.map(getNormalizedListRuntime, runtimes);
```
Comment (mlilynolting) [inline src/libs/ajax/leonardo/Runtimes.ts:246]: Again not part of this PR, but I wonder why we do the tool label normalizing in listV2 but not listV2WithWorkspace?

**Review by mlilynolting — CHANGES_REQUESTED** [review]
This looks great, and I verified it fixes the bug for me locally.
The only change I think is needed is reverting runtimeConfigBaseCost's `config` param back to a GoogleRuntimeConfig.

**Comment by mlilynolting** [comment]
I'm seeing errors in cost calculation - created a GCP workspace + Jupyter runtime, and the GcpComputeModal shows $0 for running and paused costs.

- [x] fix cost estimates

**Review by mlilynolting — APPROVED** [review]

File: src/analysis/ContextBar.test.ts (hunk 8 above)
Comment (mlilynolting) [reply src/analysis/ContextBar.test.ts:243]: (nit; not doing)

File: src/analysis/utils/cost-utils.ts [hunk 12]
```diff
@@ -48,41 +47,51 @@ import { cloudProviderTypes } from 'src/libs/workspace-utils';
 // GOOGLE COST METHODS begin
 
 export const dataprocCost = (machineType: string, numInstances: number): number => {
+  // console.log('dataprocCost', { machineType, numInstances });
```
Comment (mlilynolting) [inline src/analysis/utils/cost-utils.ts:50]: uncommenting these will let us follow the cost logic - testing of this area is sparse and does not reliably catch defects

**Review by LizBaldo — APPROVED** [review]
